lib/internal/fs/glob.js

```diff
@@ -0,0 +1,370 @@
+'use strict';
+const { lstatSync, readdirSync } = require('fs');
+const { join, resolve } = require('path');
+
+const {
+  kEmptyObject,
+} = require('internal/util');
+const {
+  validateFunction,
+  validateObject,
+} = require('internal/validators');
+
+const {
+  ArrayFrom,
+  ArrayPrototypeMap,
+  ArrayPrototypeFlatMap,
+  ArrayPrototypePop,
+  ArrayPrototypePush,
+  ArrayPrototypeSome,
+  SafeMap,
+  SafeSet,
+  StringPrototypeEndsWith,
+} = primordials;
+
+let minimatch;
+function lazyMinimatch() {
+  minimatch ??= require('internal/deps/minimatch/index');
+  return minimatch;
+}
+
+const isWindows = process.platform === 'win32';
+const isOSX = process.platform === 'darwin';
+
+class Cache {
+  #cache = new SafeMap();
+  #statsCache = new SafeMap();
+  #readdirCache = new SafeMap();
+
+  statSync(path) {
+    if (this.#statsCache.has(path)) {
+      return this.#statsCache.get(path);
+    }
+    let val;
+    try {
+      val = lstatSync(path);
+    } catch {
+      val = null;
+    }
+    this.#statsCache.set(path, val);
+    return val;
+  }
+  addToStatCache(path, val) {
+    this.#statsCache.set(path, val);
+  }
+  readdirSync(path) {
+    if (this.#readdirCache.has(path)) {
+      return this.#readdirCache.get(path);
+    }
+    let val;
+    try {
+      val = readdirSync(path, { __proto__: null, withFileTypes: true });
+    } catch {
+      val = [];
+    }
+    this.#readdirCache.set(path, val);
+    return val;
+  }
+  add(path, pattern) {
+    let cache;
+    if (this.#cache.has(path)) {
+      cache = this.#cache.get(path);
+    } else {
+      cache = new SafeSet();
+      this.#cache.set(path, cache);
+    }
+    const originalSize = cache.size;
+    pattern.indexes.forEach((index) => cache.add(pattern.cacheKey(index)));
+    return cache.size !== originalSize + pattern.indexes.size;
+  }
+  seen(path, pattern, index) {
+    return this.#cache.get(path)?.has(pattern.cacheKey(index));
+  }
+}
+
+class Pattern {
+  #pattern;
+  #globStrings;
+  indexes;
+  symlinks;
+  last;
+
+  constructor(pattern, globStrings, indexes, symlinks) {
+    this.#pattern = pattern;
+    this.#globStrings = globStrings;
+    this.indexes = indexes;
+    this.symlinks = symlinks;
+    this.last = pattern.length - 1;
+  }
+
+  isLast(isDirectory) {
+    return this.indexes.has(this.last) ||
+      (this.#pattern[this.last] === '' && isDirectory &&
+      this.indexes.has(this.last - 1) && this.#pattern[this.last - 1] === lazyMinimatch().GLOBSTAR);
+  }
+  isFirst() {
+    return this.indexes.has(0);
+  }
+  get hasSeenSymlinks() {
+    return ArrayPrototypeSome(ArrayFrom(this.indexes), (i) => !this.symlinks.has(i));
+  }
+  at(index) {
+    return this.#pattern[index];
```

Comment:
```suggestion
    return this.#pattern.at(index);
```

?

Reply:
```suggestion
    return ArrayPrototypeAt(this.#pattern, index);
```
if you’re going this route